### ISSUE TITLE
Erratic behaviour for automatic refreshes in Eclipse

### ISSUE THREAD

**Issue by kriegfrj** [bndtools/bnd#3386]
While using the continuous testing feature in #3145 in my own development, I've noticed a couple of strange things that happen from time-to-time that may or may not be related:

1. *Running instance disconnects*: sometimes (I think when the launcher is doing a refresh), it seems that the running OSGi instance becomes disconnected from the Eclipse host:
   1. In Task Manager I can see that the running instance is still there, but it is a top-level process rather than a child of the Eclipse process.
   2. The Eclipse console window associated with the OSGi instance disappears (I have it configured so that normally when a process exits, the console remains open with the <terminated> descriptor associated with it).
   3. Bndtools allows me to re-launch the same application without warning (normally if you try and launch it while there is an instance already running, it will warn you).
2. *Resource change listener stops listening*: occasionally it seems that the resource change listener stops responding to changes in built bundles and stops re-loading them.

As I noted, this is while using #3145. I'm not sure if it is happening on master or if it's a bug I introduced in #3145. I'm also not able to reliably reproduce it at this stage. I thought I'd raise this issue in case someone else had seen the same or similar behaviour and had more data to add.

Intuitively, I feel that this might have something to do with concurrency. Anecdotally, I've noticed some strange behaviour when (eg) makes some changes to a bundle that is lower down in the dependency chain - this has a ripple-through effect as several other bundles get rebuilt, which causes the resource change listener in the `AbstractOSGiLaunchDelegate` to fire off several calls to update() on the running instance in quick succession. At a first glance, it doesn't appear that there is any synchronization or concurrency handling around access to the `launcher.properties` file which is used to communicate the configuration between the host and the running OSGi instance. So it seems quite possible that there could be two concurrent calls to `ProjectLauncherImpl.writeProperties()` that interfere with each other and produce a corrupted `launcher.properties` file, or that a call to `writeProperties()` might happen concurrently with the call to loadProperties in `Launcher.watch()` and the read gets corrupted. This could cause the erratic behaviour that I've seen.

**Comment by kriegfrj** [comment]
I've implemented file locking around `launch.properties` in my local fork to see if it would fix this problem. It seems that it may have fixed the second issue but it hasn't fixed the first.

@bjhargrave suggested on the mailing list that the "detached console" phenomenon is an Eclipse bug as he has seen something similar happen with regular JUnit tests. In which case the bug needs to be reported there rather than here.

I'll try and tidy up my file locking patch and submit it for consideration as a solution to the second issue.

**Comment by kriegfrj** [comment]
Of the two issues raised in the OP:

* I think that we have established that the first of these issues in an Eclipse issue and not directly under our control.
* The remaining issue (possibly related to file locking) seems also to have disappeared. I suspect that it was resolved (for practical purposes) by #3627 - now that the launcher waits for the entire build to finish before updating the properties file, there is a vanishingly small chance that there might be two simultaneous updates happening at the same time.  In theory file locking on updates the the properties file might still be a good idea, but until there is an imminent need I'm going to spend any more time on it.
